wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -0,0 +1,578 @@
+import org.eclipse.che.api.core.notification.EventService;
+import org.eclipse.che.api.machine.server.MachineManager;
+import org.eclipse.che.api.machine.server.exception.MachineException;
+import org.eclipse.che.api.machine.server.exception.SnapshotException;
+import org.eclipse.che.api.machine.server.model.impl.MachineConfigImpl;
+import org.eclipse.che.api.machine.server.model.impl.MachineImpl;
+import org.eclipse.che.api.workspace.server.model.impl.EnvironmentImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceRuntimeImpl;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent;
+import org.eclipse.che.api.workspace.shared.dto.event.WorkspaceStatusEvent.EventType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import static java.lang.String.format;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+
+/**
+ * Defines an internal API for managing {@link WorkspaceRuntimeImpl} instances.
+ *
+ * <p>This component implements {@link WorkspaceStatus} spec.
+ *
+ * <p>All the operations performed by this component are synchronous.
+ *
+ * <p>The implementation is thread-safe and guarded by {@link ReentrantReadWriteLock rwLock}.
+ *
+ * <p>The implementation doesn't validate parameters.
+ * Parameters should be validated by caller of methods of this class.
+ *
+ * @author Yevhenii Voevodin
+ * @author Alexander Garagatyi
+ */
+@Singleton
+public class WorkspaceRuntimes {
+
+    private static final Logger LOG = LoggerFactory.getLogger(WorkspaceRuntimes.class);
+
+    private final ReadWriteLock                         rwLock;
+    private final Map<String, RuntimeDescriptor>        descriptors;
+    private final Map<String, Queue<MachineConfigImpl>> startQueues;
+    private final MachineManager                        machineManager;
+    private final EventService                          eventService;
+
+    private volatile boolean isPreDestroyInvoked;
+
+    @Inject
+    public WorkspaceRuntimes(MachineManager machineManager, EventService eventService) {
+        this.machineManager = machineManager;
+        this.eventService = eventService;
+        this.descriptors = new HashMap<>();
+        this.startQueues = new HashMap<>();
+        this.rwLock = new ReentrantReadWriteLock();
+    }
+
+    /**
+     * Returns the runtime descriptor describing currently starting/running/stopping
+     * workspace runtime, throws {@link NotFoundException} if workspace with
```

Comment:
I don't think you should explain the whole contract in the first sentence. Let @throws annotation describe when exception is thrown.